README.rst
=======
Caravan
=======

**Light python framework for AWS SWF**

About this Project
==================

Caravan is being used at `Ludia <https://github.com/ludia>`_ for projects
like marketing campaign system and on-demand distributed processing
systems (thanks to hundreds of Lambda functions).

Feedbacks, ideas and contributions are highly welcomed. (Just open a
`Github issue <https://github.com/pior/caravan/issues>`_).

- `Code on Github <https://github.com/pior/caravan>`_
- `PyPI <https://pypi.python.org/pypi/caravan>`_
- `Tests <https://travis-ci.org/pior/caravan>`_ |travis| |coveralls|
- Doc: ``TODO``

.. |travis| image:: https://travis-ci.org/pior/caravan.svg?branch=master
   :target: https://travis-ci.org/pior/caravan

.. |coveralls| image:: https://coveralls.io/repos/pior/caravan/badge.svg?branch=master&service=github
   :target: https://coveralls.io/github/pior/caravan?branch=master

Focus of this project
=====================

Similar projects exists (like Simpleflow which seems mature).
Here is where Caravan differs from existing projects:

- Support of AWS Lambda tasks
- Boto3
- KISS
- Bring your own workflow framework (standard implementations as contribs)
- No coupling between Decider code and Activity code
- Paster compatible config file

Features
========

- Decider worker
- Activity task worker
- Commands to start/signal/terminate an arbitrary workflow execution
- Command to list open workflow execution
- Command to register a domain / list domains

Configuration
=============

Caravan uses Boto3 to connect to AWS. See
`Boto 3 configuration guide <http://boto3.readthedocs.org/en/latest/guide/configuration.html>`_
for the complete documentation.

.. warning::
   On AWS EC2, the metadata provider only provides the credentials, the
   must be provided by configuration.

Environment Variables
---------------------

``AWS_ACCESS_KEY_ID``
    The access key for your AWS account.

``AWS_SECRET_ACCESS_KEY``
    The secret key for your AWS account.

``AWS_DEFAULT_REGION``
    The default region to use, e.g. `us-east-1`.

``AWS_PROFILE``
    The default credential and configuration profile to use, if any.

Configuration Files
-------------------

The credentials file is located at ``~/.aws/credentials``::

   [default]
   # The access key for your AWS account
   aws_access_key_id=<YOUR ACCESS KEY ID>

   # The secret key for your AWS account
   aws_secret_access_key=<YOUR SECRET KEY>

The settings file is located at ``~/.aws/config``::

   [default]
   # The default region when making requests
   region=<REGION NAME>

It also supports profiles::

   [profile dev-profile]
   # The default region when using the dev-profile account
   region=<REGION NAME>

Demo
====

Setup a SWF domain to run this example::

   $ caravan-domain-register -n CaravanDemo --retention-days 1

Write a workflow type (see full demo_)

.. code:: python

   from caravan import Workflow


   class Demo(Workflow):

       """Noop workflow using the bare caravan API."""

       name = 'Demo'
       version = '0.1'
       default_execution_start_to_close_timeout = '600'
       default_task_start_to_close_timeout = '10'

       def run(self):
           self.task.print_events()
           self.task.add_decision('CompleteWorkflowExecution')

.. _demo: https://github.com/pior/caravan/blob/master/caravan/examples/demo.py

Run the decider with the Demo workflow::

   $ caravan-decider -d CaravanDemo -m caravan.examples.demo -t default --verbose

Start an execution of the Demo workflow::

   $ caravan-start -d CaravanDemo -n Demo -v 0.1 -i 1

   (The Demo workflow will wait for 5 minutes)

List the executions::

   $ caravan-list -d CaravanDemo
   $ caravan-list -d CaravanDemo --oldest 2015-01-01

Send a signal to an execution::

   $ caravan-signal -d CaravanDemo -i 1 -s PRINT --input 'Hello World!'
   $ caravan-signal -d CaravanDemo -i 1 -s PRINT --input 'Lorem ipsum'
   $ caravan-signal -d CaravanDemo -i 1 -s STOP

Terminate an execution::

   $ caravan-terminate -d CaravanDemo -i 1

Similar projects
================

Python:

- Simpleflow: https://github.com/botify-labs/simpleflow
- Flowy: https://github.com/severb/flowy
- Garcon: https://github.com/xethorn/garcon

Ruby:

- AWS Flow: https://github.com/aws/aws-flow-ruby

References
==========

- `AWS Developer Guide for SWF <http://docs.aws.amazon.com/amazonswf/latest/developerguide/>`_
- `AWS API reference for SWF <http://docs.aws.amazon.com/amazonswf/latest/apireference/>`_

Development
===========

Possibly use `virtualenvwrapper <https://virtualenvwrapper.readthedocs.org/en/latest/>`_
to manage your virtualenvs.

   $ mkvirtualenv caravan

   or

   $ virtualenv caravan

Clone and install development dependencies::

   (caravan)$ git clone git@github.com:pior/caravan.git
   (caravan)$ cd caravan
   (caravan)$ pip install -e .[dev]

Run tests::

   (caravan)$ nosetests

Release
=======

The release process use zest.releaser::

   $ fullrelease

License
=======

MIT licensed. See the bundled
`LICENSE <https://github.com/pior/caravan/blob/master/LICENSE>`_
file for more details


CHANGES.rst
=========
Changelog
=========

0.2.3 (unreleased)
------------------

- Nothing changed yet.


0.2.2 (2015-12-31)
------------------

- Add a workflow demo (caravan.examples.demo)
- Changed Workflow api for defaults
- Support workflow default timeouts as type int


0.2.1 (2015-12-07)
------------------

- Add command caravan-domain-list
- Mock sys.argv rather than using a special argument


0.2.0 (2015-12-03)
------------------

- Switch to unittest + nose for tests
- Tested with Python 2.7, 3.3, 3.4, 3.5
- Don't force the AWS region to us-east-1


0.1.4 (2015-12-02)
------------------

- Allow workflows to raise an exception to fail an execution


0.1.3 (2015-11-24)
------------------

- Release again to fix PyPI upload server error


0.1.2 (2015-11-24)
------------------

- Fix crash with --logging-config


0.1.1 (2015-11-23)
------------------

- Add command caravan-terminate
- Make --modules as config key works


0.1.0 (2015-11-19)
------------------

- Add command caravan-domain-register
- Add command caravan-list


0.0.2 (2015-11-12)
------------------

- First release on Pypi


0.0.1 (unreleased)
------------------

- Initial setup


caravan/examples/demo.py
from __future__ import print_function
import time

from caravan import Workflow, Activity


class DemoWorkflow(Workflow):

    """Demo workflow using the bare caravan API."""

    name = 'Demo'
    version = '0.1'
    default_execution_start_to_close_timeout = 600
    default_task_start_to_close_timeout = 10
    default_task_list = 'default'
    default_child_policy = 'TERMINATE'

    def run(self):
        # For development, you can print the history events
        print("")
        self.task.print_events()

        # Minimalist History evaluation, just consider the last event
        non_decision_events = [e for e in self.task.events
                               if not e['eventType'].startswith('Decision')]
        last_event = non_decision_events[-1]

        if last_event['eventType'] == 'WorkflowExecutionStarted':
            # This is the start of this execution !
            # First, let's wait for a signal using a timer
            self.task.add_decision('StartTimer',
                                   startToFireTimeout='300',
                                   timerId='SignalTimeout')
            self.task.decision_done(msg='JustStartedATimer')

        elif last_event['eventType'] == 'TimerFired':
            # The timer fired! Timeout!
            # We fail the whole workflow execution (while providing a reason)
            self.task.fail('patience_lost')

        elif last_event['eventType'] == 'WorkflowExecutionSignaled':
            # Hey! We received a Signal!

            signal_name = last_event['attributes']['signalName'].upper()
            # signal_input = last_event['input']

            if signal_name == 'ABORT':
                # Fail the workflow execution right now
                self.task.fail('aborted')

            elif signal_name == 'PRINT':
                # Let's print the Signal input (payload)
                signal_input = last_event['attributes'].get('input')
                print("\nThe signal input is: '%s'" % signal_input)

            elif signal_name == 'STOP':
                # Mark this workflow execution as complete
                workflow_result = '{"success": true}'
                self.task.add_decision('CompleteWorkflowExecution',
                                       result=workflow_result)

            elif signal_name == 'ACTIVITY':
                # Schedule a demo activity
                activity_type = {
                    'name': DemoActivity.name,
                    'version': DemoActivity.version,
                    }
                activity_id = str(time.time())
                self.task.add_decision('ScheduleActivityTask',
                                       activityType=activity_type,
                                       activityId=activity_id,
                                       input='"DemoInput"')

            else:
                print("Unknown signal '%s'" % signal_name)
                self.task.decision_done(msg='JustIgnoringThisUnknownSignal')

        else:
            print('Unknown last event type')


class DemoActivity(Activity):

    name = 'DemoActivity'
    version = '0.1.2'

    default_task_list = 'default'

    default_task_start_to_close_timeout = 60
    default_task_schedule_to_start_timeout = 10
    default_task_schedule_to_close_timeout = 70
    default_task_heartbeat_timeout = 'NONE'

    def run(self, input):
        print('Received input: %s' % input)
        self.do_stuff()
        return {'input': input, 'output': 'YO'}

    def do_stuff(self):
        print('Doing stuff (waiting 3 secs)')
        time.sleep(2)
